flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -252,18 +249,27 @@ public void failJobIfNotEnoughResources() throws Exception {
 			findThrowableWithMessage(
 				failureCause,
 				"Could not allocate the required slot within slot request timeout.").isPresent());
+		assertThat(jobStatus, is(equalTo(JobStatus.FAILED)));
 	}
 
-	private void drainAllAvailableSlots() {
-		final int numberOfAvailableSlots = slotProvider.getNumberOfAvailableSlots();
-		for (int i = 0; i < numberOfAvailableSlots; i++) {
-			slotProvider.allocateSlot(
-				new SlotRequestId(),
-				new ScheduledUnit(new JobVertexID(), null, null),
-				SlotProfile.noRequirements(),
-				true,
-				Time.milliseconds(TIMEOUT_MS));
-		}
+	@Test
+	public void skipDeploymentIfVertexVersionOutdated() {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+
+		final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+		final List<ExecutionVertexID> initiallyScheduledVertices = testExecutionVertexOperations.getDeployedVertices();
+
+		final ArchivedExecutionVertex onlyExecutionVertex = Iterables.getOnlyElement(scheduler.requestJob().getAllExecutionVertices());
+		final ExecutionAttemptID attemptId = onlyExecutionVertex.getCurrentExecutionAttempt().getAttemptId();
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.FAILED));
+
+		testExecutionSlotAllocator.disableAutoCompletePendingRequests();
+		taskRestartExecutor.triggerScheduledTasks();
+		executionVertexVersioner.recordModification(new ExecutionVertexID(getOnlyJobVertex(jobGraph).getID(), 0));
+		testExecutionSlotAllocator.completePendingRequests();
+
+		assertThat(initiallyScheduledVertices, is(equalTo(testExecutionVertexOperations.getDeployedVertices())));
```

Comment:
One more thing we may need to verify is that concurrent failovers should not result in more failovers.
We can do it by limit the max restart attempts or check the final attempt number of a vertex.

Reply:
Fixed, I think.